### DISCUSSION TITLE
Sampler: dynamic sample rate

### DISCUSSION THREAD

**Discussion by shawnsong** [open-telemetry/opentelemetry-java#4504] [Q&A]
### Description

Our production systems recently experienced some issues caused by the `TraceIdRatioBasedSampler`. We use our own flag system to dynamically update the sample rate in our production. Our current way of changing the sample rate dynamically is by having a proxy class and re-initializing the real sampler inside the proxy class. The issue happened when an invalid sample rate was set in the flag system and exceptions were thrown when calling `Sampler.traceIdRatioBased()`. My questions are:

1. Is there a recommended/native way to update the sample rate dynamically?
2. Is it worth making the above-mentioned method throws a checked exception so it won't be overlooked? I know this will probably bring in the checked/unchecked exception argument but the reason for this is given the code is considered as _initialization_ so even if it throws a checked exception, we won't end up with `try-catch` blocks everywhere in the codebase.

**Comment by anuraaga** [comment]
Hi @shawnsong - sorry to hear about the production issue.

1. Wrapping the `Sampler` is indeed the way to make it updatable. We use a similar technique in zpages https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/zpages/src/main/java/io/opentelemetry/sdk/extension/zpages/TracezTraceConfigSupplier.java#L37
2. Most of our factories don't throw checked exceptions even when they have validation. While there are probably pros and cons of each approach, our semantic versioning policy means we wouldn't be able to add to the existing method for now even if we wanted to, as adding a checked exception would cause existing code that does compile to stop compiling when updating their dependency.

**Comment by shawnsong** [comment]
Hi @anuraaga, 

Thanks for the prompt response. 

> Wrapping the Sampler is indeed the way to make it updatable.

Is this going to be the long-term solution or it's a workaround for now. Are there any plans to introduce the dynamic sample rate feature?

> Most of our factories don't throw checked exceptions even when they have validation...

I can understand from the semantic versioning perspective. Would that be possible in the future? The [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md#basic-error-handling-principles) is slightly vague in this area as what is considered as runtime vs configuration. Is it worth adding another method like `Sampler.traceIdRatioBasedOrDefault()` for the runtime so if any invalid config is used, it just defaults to a `NOOP` or something that's guaranteed not to fail?

**Comment by codeforaliving** [comment]
@shawnsong I am trying to do the same thing.  We want the ability to update the trace sample rate/ratio at runtime.  Did you end finding a working solution?  Would you mind sharing it?

**Reply by shawnsong** [reply]
Hi @codeforaliving,

You can have a custom sampler that implements `io.opentelemetry.sdk.trace.samplers.Sampler`. A simplified version can be something like:

```
public class TraceSampler implements Sampler {
    private volatile Sampler delegate;
    public TraceSampler() {
        Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate(() -> 
                delegate = Sampler.traceIdRatioBased(getYourSampleRate()), 0, 10, TimeUnit.SECONDS);
    }

    @Override
    public SamplingResult shouldSample(Context parentContext, String traceId, String name,
                                       SpanKind spanKind, Attributes attributes, List<LinkData> parentLinks) {
        return delegate.shouldSample(parentContext, traceId, name, spanKind, attributes, parentLinks);
    }

    @Override
    public String getDescription() {
        return "A customized sampler";
    }
}

```

Please note that `Sampler.traceIdRatioBased()` would throw `IllegalArgumentException` if an invalid sample rate is provided.

**Comment by codeforaliving** [comment]
@shawnsong thank you for sharing the code. Would this code live in the extension or in the app?

**Reply by shawnsong** [reply]
I don't quite get the question. We have it configured for our tracer provider
```
SdkTracerProvider.builder()
  .addSpanProcessor(..)
  .setSampler(new TraceSampler())
  .setIdGenerator(...)
  .build();
```
